### PR TITLE
Feature/controllers and tests

### PR DESCRIPTION
**Pull request by ajax-makarytskyi-o**
Controllers, tests and detekt plugin were added. Services, exceptions, dto and repositories were changed.

### CODE REVIEW

File: src/main/kotlin/com/makarytskyi/rentcar/exception/ResourceNotFoundException.kt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package com.makarytskyi.rentcar.exception
+
+class ResourceNotFoundException(message: String) : Exception(message)
```
Comment (ajax-butriy-v) [inline src/main/kotlin/com/makarytskyi/rentcar/exception/ResourceNotFoundException.kt]:
```suggestion
class ResourceNotFoundException(message: String) : RuntimeException(message)
```

File: src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.PutMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.ResponseStatus
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/api/v1/car")
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt]: needed -> /api/v1/cars

File: src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+class CarController(private val service: CarService) {
+
+    @GetMapping("/{id}")
+    fun findById(@PathVariable id: String): CarResponse = service.getById(id)
+
+    @GetMapping("/all")
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt]: Just @GetMapping()

File: src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        val errorResponse = ErrorResponse(
+            LocalDateTime.now(),
+            HttpStatus.BAD_REQUEST.value(),
+            ex.message ?: "Argument is invalid"
+        )
+        return ResponseEntity(errorResponse, HttpStatus.BAD_REQUEST)
+    }
+}
+
+data class ErrorResponse(
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt]: Change visibility of the access modifier

File: src/main/kotlin/com/makarytskyi/rentcar/controller/RepairingController.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    @PutMapping("/{id}")
+    fun update(@PathVariable id: String, @Valid @RequestBody repairing: UpdateRepairingRequest): RepairingResponse =
+        service.update(id, repairing)
+
+    @GetMapping("/car/{carId}")
+    fun findByCarId(@PathVariable carId: String): List<RepairingResponse> = service.findByCarId(carId)
+
+    @GetMapping("/status/{status}")
+    fun findByStatus(@PathVariable status: Repairing.RepairingStatus): List<RepairingResponse> =
```
Comment (ajax-butriy-v) [inline src/main/kotlin/com/makarytskyi/rentcar/controller/RepairingController.kt]:
```suggestion
    @GetMapping
    fun findByStatus(@RequestParam status: Repairing.RepairingStatus): List<RepairingResponse> =
```

same comment for other endpoints with same pattern

File: src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryOrderRepository.kt [hunk 6]
```diff
@@ -25,7 +25,8 @@ class InMemoryOrderRepository : OrderRepository {
         map.remove(id)
     }
 
-    override fun findAllByDate(date: Date): List<Order> = map.values.filter { it.from != null && it.to != null }.filter { date.before(it.to) && date.after(it.from) }
+    override fun findAllByDate(date: Date): List<Order> =
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryOrderRepository.kt]: Too long function to read. Prefer to write code from Top to Bottom

File: src/test/kotlin/services/CarServiceTests.kt [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    fun `update should return updated car`() {
+        //GIVEN
+        whenever(carRepository.update(carId, updateCarEntity)).thenReturn(updatedCar)
+
+        //WHEN
+        val result = carService.update(carId, updateCarRequest)
+
+        //THEN
+        assertNotNull(result)
+        assertEquals(250, result.price)
```
Comment (ajax-malov-n) [inline src/test/kotlin/services/CarServiceTests.kt]: Magic number

File: src/testFixtures/kotlin/CarFixture.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
+import com.makarytskyi.rentcar.dto.car.CarResponse
+import com.makarytskyi.rentcar.dto.car.CreateCarRequest
+import com.makarytskyi.rentcar.dto.car.UpdateCarRequest
+import com.makarytskyi.rentcar.model.Car
+
+object CarFixture {
```
Comment (ajax-malov-n) [inline src/testFixtures/kotlin/CarFixture.kt]: Make a folder for test fixtures

File: src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.ResponseStatus
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/cars")
+class CarController(private val service: CarService) {
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/controller/CarController.kt]: make every class as internal by default

File: src/main/kotlin/com/makarytskyi/rentcar/controller/OrderController.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+@RestController
+@RequestMapping("/orders")
+class OrderController(private val service: OrderService) {
+
+    @GetMapping("/{id}")
+    fun findById(@PathVariable id: String): OrderResponse = service.getById(id)
+
+    @GetMapping()
+    fun findAll(
+        @RequestParam(required = false) carId: String?,
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/controller/OrderController.kt]: separate to multiple API with path variable per user and per car

- /orders/user/{userId}
- /orders/car/{carId}
- /orders/user/{userId}/car/{carId} or /orders/car/{carId}/user/{userId}
- /orders

Same is true in other controllers

File: src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.RestControllerAdvice
+
+@RestControllerAdvice
+class GlobalExceptionHandler {
+    @ExceptionHandler(ResourceNotFoundException::class)
+    internal fun handleResourceNotFoundException(ex: ResourceNotFoundException): ResponseEntity<ErrorResponse> {
+        val errorResponse = ErrorResponse(
+            LocalDateTime.now(),
+            HttpStatus.NOT_FOUND.value(),
+            ex.message ?: "Resource is not found"
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt]: HttpStatus.NOT_FOUND is enough. Message is ambiguous.

File: src/testFixtures/kotlin/fixtures/UserFixture.kt [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+import com.makarytskyi.rentcar.dto.user.CreateUserRequest
+import com.makarytskyi.rentcar.dto.user.UpdateUserRequest
+import com.makarytskyi.rentcar.dto.user.UserResponse
+import com.makarytskyi.rentcar.model.User
+
+object UserFixture {
+    const val userId: String = "1241242"
+    const val createdUserId: String = "83465294"
+
+    val existingUser = User(
```
Comment (ajax-samondros-b) [inline src/testFixtures/kotlin/fixtures/UserFixture.kt]: make Fixture to always return new POJO with random values. You can specify value only when it is desired by the Flow, e.g some enum.

File: src/test/kotlin/services/UserServiceTests.kt [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    @Test
+    fun `findAll should return empty List of UserResponse if repository return empty List`() {
+        //GIVEN
+        whenever(userRepository.findAll()).thenReturn(emptyList())
+
+        //WHEN
+        val result = userService.findAll()
+
+        //THEN
+        verify(userRepository).findAll()
```
Comment (ajax-samondros-b) [inline src/test/kotlin/services/UserServiceTests.kt]: asserts in the beginning

File: src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.ExceptionHandler
+import org.springframework.web.bind.annotation.RestControllerAdvice
+
+@RestControllerAdvice
+internal class GlobalExceptionHandler {
+    @ExceptionHandler(ResourceNotFoundException::class)
+    internal fun handleResourceNotFoundException(): ResponseEntity<ErrorResponse> {
+        val errorResponse = ErrorResponse(
+            LocalDateTime.now(),
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt]: why this timestamp is needed?

Comment (ajax-makarytskyi-o) [reply src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt]: to get time when exception happened

File: src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryCarRepository.kt [hunk 15]
```diff
@@ -5,9 +5,10 @@ import com.makarytskyi.rentcar.repository.CarRepository
 import org.bson.types.ObjectId
 import org.springframework.stereotype.Repository
 
+
 @Repository
-class InMemoryCarRepository: CarRepository {
-    val map: MutableMap<String, Car> = HashMap()
+class InMemoryCarRepository : CarRepository {
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryCarRepository.kt]: internal?

File: src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryOrderRepository.kt [hunk 16]
```diff
@@ -8,7 +8,7 @@ import org.springframework.stereotype.Repository
 
 @Repository
 class InMemoryOrderRepository : OrderRepository {
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryOrderRepository.kt]: internal

File: src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryRepairingRepository.kt [hunk 17]
```diff
@@ -8,7 +8,7 @@ import org.springframework.stereotype.Repository
 
 @Repository
 class InMemoryRepairingRepository: RepairingRepository {
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryRepairingRepository.kt]: internal

File: src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryCarRepository.kt (hunk 15 above)
Comment (ajax-makarytskyi-o) [reply src/main/kotlin/com/makarytskyi/rentcar/repository/impl/InMemoryCarRepository.kt]: Yes, I missed repositories. I will fix this

File: src/main/kotlin/com/makarytskyi/rentcar/service/OrderService.kt [hunk 18]
```diff
@@ -3,84 +3,84 @@ package com.makarytskyi.rentcar.service
...
 
-    fun findByCarId(carId: String): List<OrderResponse> = orderRepository.findByCarId(carId).map {
+    fun findByCar(carId: String): List<OrderResponse> = orderRepository.findByCarId(carId).map {
         Order.toResponse(it, getCarPrice(it.carId))
     }
 
-    fun update(id: String, orderRequest: UpdateOrderRequest): OrderResponse {
+    fun findByCarAndUser(carId: String, userId: String): List<OrderResponse> =
+        orderRepository.findByUserIdAndCarId(carId, userId)
+            .map { Order.toResponse(it, getCarPrice(it.carId)) }
```
Comment (ajax-samondros-b) [inline src/main/kotlin/com/makarytskyi/rentcar/service/OrderService.kt]: Domain should know nothing about response. Please invert. Call `OrderResponse.from(Order)`

**Review by ajax-samondros-b — CHANGES_REQUESTED** [review]
see comments

File: src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt (hunk 14 above)
Comment (ajax-samondros-b) [reply src/main/kotlin/com/makarytskyi/rentcar/exception/GlobalExceptionHandler.kt]: do you use it somehow? if not please clean it out

**Review by ajax-samondros-b — APPROVED** [review]

**Pull request merged**